### PR TITLE
fix: Relative URLS supports & If not a URL, pass #417

### PR DESCRIPTION
**Pull request by black7375**
### Description

- If it starts with `//`, replace it with the same protocol.
- To avoid unsupported protocols, such as `mailto` and `tel`, we only allow writing `http` and `https`.

### Issues

- Closes: #417 

### Testing

- Add a unit test for cases that start with `//`.

### Checklist

- [x] CI passed

### CODE REVIEW

**Review by vdusek — CHANGES_REQUESTED** [review]
Hey, great job on the implementation! 

I have just two suggestions:

1. **Refactoring URL logic**: To avoid repeating the same logic across multiple crawlers, we could move these functions to the `_utils/urls.py` module.

2. **URL concatenation**: I recommend using `urllib.parse.urljoin` instead of f-string concatenation.

`_utils/urls.py`:

```python
# ...

def is_url_absolute(url: str) -> bool:
    """Check if a URL is absolute."""
    # This case has to be handled explicitly.
    if url.startswith('//'):
        return False

    return bool(urlparse(url).netloc)


def is_http_url(url: str) -> bool:
    """Check if a URL is an HTTP/HTTPS."""
    return url.lower().startswith(('http://', 'https://'))


def convert_to_absolute_url(base_url: str, relative_url: str) -> str:
    """Convert a relative URL to an absolute URL using a base URL."""
    if relative_url.startswith('//'):
        protocol = urlparse(base_url).scheme
        return urljoin(protocol, relative_url)

    return urljoin(base_url, relative_url)

# ...
```    

`beautifulsoup_crawler/beautifulsoup_crawler.py`:

```python
# ...

                if (url := link.attrs.get('href')) is not None:
                    url = url.strip()

                    if not is_url_absolute(url):
                        url = convert_to_absolute_url(context.request.url, url)

                    if not is_http_url(url):
                        continue

# ...
```

**Comment by janbuchar** [comment]
Please, could we have some unit tests before we merge this? 🙂

**Comment by aka-Ani** [comment]
I agree this would "solve" the issue but it feels unnecessary to do string checks to verify things like `http` `https` or `//`. The urllib.parse library in addition to Pydantic models cover this functionality already. My suggestion would be to modify the crawlers to something more like this:
image

Essentially, relying on the urllib library to best construct a working absolute URL out of the href tags (it does this reliably for protocol relative URLs, as well as other relative URLs). And then instead of checking for specific scenarios to catch/pass, just handle any validation error Pydantic may raise. Since we are using the `AnyHttpUrl` Pydantic type, this should filter out all the unwanted URLs. 

I just think its cleaner approach vs introducing more (possibly redundant) functions into the codebase. But at the end of the day I suppose they both solve this specific scenario just fine, and I will leave it up to you guys 🙂

**Comment by black7375** [comment]
1. Added a condition to `is_url_absolute` to check for URL scheme.
2. As suggested, it'll base `continue` on `try-except`.
3. Add a unit test for cases that start with `//`.

**Comment by aka-Ani** [comment]
A couple things I noticed in the last set of changes:
1. Are you sure you should be handling a `ValueError` exception? I thought that line of code would raise a `ValidationError`
2. Maybe this is subjective, but I don't really see the purpose of checking `is_url_absolute` in these scenarios at all. The `urllib.parse.urljoin` function (reused as `convert_to_absolute_url` in utils) works just fine to construct a full/absolute url from the html extracted url and the base url. Taking the 7 examples from the url unit tests file (first 3 True, last 4 False), you can see that `urljoin` returns the same url if it was absolute or returns the joined url if it was not.
image
I think this would be the same output as doing the `if not is_url_absolute` then `convert_to_absolute_url`, but maybe it makes the code more readable or something, I'll let you decide.

**Comment by black7375** [comment]
1. I think you're right. I pasted the content from the review, but I feel that `ValidationError` is more correct.
2. I'd like to approach this a bit defensively, as unexpected side effects might occur. However, as I don't have a strong opinion, I will follow @vdusek 's opinion in the follow-up review.

**Review by vdusek — APPROVED** [review]
LGTM, thank!

**Pull request merged**
